pydantic/schema.py

```diff
@@ -0,0 +1,392 @@
+from datetime import date, datetime, time, timedelta
+from decimal import Decimal
+from enum import Enum
+from pathlib import Path
+from typing import Any, Dict, Sequence, Set, Tuple, Type
+from uuid import UUID
+
+from . import main
+from .fields import Field, Shape
+from .json import pydantic_encoder
+from .types import DSN, UUID1, UUID3, UUID4, UUID5, DirectoryPath, EmailStr, FilePath, Json, NameEmail, UrlStr
+from .utils import clean_docstring
+
+
+def get_flat_models_from_model(model: Type['main.BaseModel']) -> Set[Type['main.BaseModel']]:
```

Comment:
is `main.*` enough for mypy or ids to work out what `BaseModel` is? Surely `pydantic.BaseModel` or `pydantic.main.BaseModel` is required?

Reply:
also if we can import `BaseModel`, just use that rather than a string.

Reply:
I can't import it directly as these functions are imported and used inside of `BaseModel`. I had to use that ["trick" from PEP 484](https://www.python.org/dev/peps/pep-0484/#forward-references).

If you want I can change it to:

```Python
def get_flat_models_from_model(model: Type) -> Set[Type]:
```

But for it to work it needs the models to be Pydantic, i.e, to have `__fields__`. That's why I annotated it like that.

Let me know what do you prefer.

Reply:
you already import `main` above, so therefore surely you can use `main.BaseModel` or better import `BaseModel` from main and use it here.

Reply:
It can't be done with circular imports. `main.BaseModel` already depends on these methods, so `BaseModel` is not yet defined by the moment these methods are defined, as they are imported and used by `main.BaseModel` first, check the errors in CI: https://travis-ci.org/samuelcolvin/pydantic/jobs/457945798

From [PEP 484: Forward references](https://www.python.org/dev/peps/pep-0484/#forward-references):

> When a type hint contains names that have not been defined yet, that definition may be expressed as a string literal, to be resolved later.

> A situation where this occurs commonly is the definition of a container class, where the class being defined occurs in the signature of some of the methods. For example, the following code (the start of a simple binary tree implementation) does not work:

```Python
class Tree:
    def __init__(self, left: Tree, right: Tree):
        self.left = left
        self.right = right
```

> To address this, we write:

```Python
class Tree:
    def __init__(self, left: 'Tree', right: 'Tree'):
        self.left = left
        self.right = right
```

Later in the same section:

> [...] The solution is to switch to module-only imports and reference the models by their _module_._class_ name:

```Python
# File models/a.py
from models import b
class A(Model):
    def foo(self, b: 'b.B'): ...
```